Support to transfer parameter within quotation mark in global or local engine setting

jameszhouyi opened the pull request:
For current implementation, it don't support to transfer parameter  within quotation mark in "BINARY_PARAMS" since linux shell. For example, there is "--conf" option to specify parameter within double quotation mark, which the parameter include multiple value separated by white space(--conf "spark.executor.extraJavaOptions=-XX:NewRatio=1  -XX:SurvivorRatio=1") like below, it will error out "Unrecognized option: -XX:SurvivorRatio=1" when run Bigbench

Posted complete example:
BINARY_PARAMS='-v --driver-memory 4g --executor-memory 100g  --executor-cores 16 --num-executors 8 --conf  spark.yarn.executor.memoryOverhead=8192 --conf "spark.executor.extraJavaOptions=-XX:NewRatio=1  -XX:SurvivorRatio=1" --master yarn-client --jars /usr/lib/hive/lib/hive-common.jar,/usr/lib/hive/lib/hive-cli.jar,/usr/lib/hive/lib/hive-exec.jar,/usr/lib/hive/lib/hive-service.jar,/usr/lib/hive/lib/hive-metastore.jar,/usr/lib/hive/lib/libfb303-0.9.0.jar,/usr/lib/hive/lib/jdo-api-3.0.1.jar,/usr/lib/hive/lib/antlr-runtime-3.4.jar,/usr/lib/hive/lib/datanucleus-api-jdo-3.2.6.jar,/usr/lib/hive/lib/datanucleus-core-3.2.10.jar,/usr/lib/hive/lib/datanucleus-rdbms-3.2.9.jar,/usr/lib/hive/lib/derby-10.10.1.1.jar,/etc/hive/conf/hive-site.xml,/opt/Big-Bench/engines/hive/queries/Resources/bigbenchqueriesmr.jar,/opt/Big-Bench/engines/hive/queries/Resources/opennlp-maxent-3.0.3.jar,/opt/Big-Bench/engines/hive/queries/Resources/opennlp-tools-1.5.3.jar'

Manuel-Danisch commented:
Hi,

I'd like to avoid adding pipe solutions as much as possible. Pipes as well as xargs may introduce other problems. Does it work if you add "eval" before the command?

I wrote this little test bash script, and here it works as expected:

```bash
#!/usr/bin/env bash

runCmd () {
  for param in "$@"
  do
    echo "$param"
  done
}

BINARY_PARAMS='-v --driver-memory 4g --executor-memory 100g --executor-cores 16 --num-executors 8 --conf spark.yarn.executor.memoryOverhead=8192 --conf "spark.executor.extraJavaOptions=-XX:NewRatio=1 -XX:SurvivorRatio=1" --master yarn-client --jars /usr/lib/hive/lib/hive-common.jar,/usr/lib/hive/lib/hive-cli.jar,/usr/lib/hive/lib/hive-exec.jar,/usr/lib/hive/lib/hive-service.jar,/usr/lib/hive/lib/hive-metastore.jar,/usr/lib/hive/lib/libfb303-0.9.0.jar,/usr/lib/hive/lib/jdo-api-3.0.1.jar,/usr/lib/hive/lib/antlr-runtime-3.4.jar,/usr/lib/hive/lib/datanucleus-api-jdo-3.2.6.jar,/usr/lib/hive/lib/datanucleus-core-3.2.10.jar,/usr/lib/hive/lib/datanucleus-rdbms-3.2.9.jar,/usr/lib/hive/lib/derby-10.10.1.1.jar,/etc/hive/conf/hive-site.xml,/opt/Big-Bench/engines/hive/queries/Resources/bigbenchqueriesmr.jar,/opt/Big-Bench/engines/hive/queries/Resources/opennlp-maxent-3.0.3.jar,/opt/Big-Bench/engines/hive/queries/Resources/opennlp-tools-1.5.3.jar'

runCmd $BINARY_PARAMS #wrong output
echo
eval runCmd $BINARY_PARAMS #correct output
```

So, do you still have the problem if you use this command line:
```bash
eval "$BINARY" $BINARY_PARAMS $INIT_PARAMS "$@"
```

Best regards,

Manuel

jameszhouyi commented:
@Manuel-Danisch Thanks a lot Manuel ! It's OK for me with eval "$BINARY" $BINARY_PARAMS $INIT_PARAMS "$@" command line

Manuel-Danisch commented:
Hi,

pull request merged. Thanks for the patch.

Best regards,

Manuel

joshuayao commented:
The fix causes exceptions in cleanQueries. Need to be fixed.

Manuel-Danisch commented:
Hi,

which exceptions do you get with cleanQueries?

Best regards,

Manuel

jameszhouyi commented:
Hi,
Run "./bin/bigBench cleanQuery -q 1" and  it will throw below errors.This error caused by script in engine/hive/queries/q{01..30}/ like below:

query_run_clean_method () {
        runCmdWithErrorCheck runEngineCmd -e "DROP TABLE IF EXISTS $TEMP_TABLE; DROP TABLE IF EXISTS $RESULT_TABLE;"
        return $?
}



 ERROR Logs:
......
============================
</settings from hiveSettings.sql>
============================
NoViableAltException(-1@[651:1: ddlStatement : ( createDatabaseStatement | switchDatabaseStatement | dropDatabaseStatement | createTableStatement | dropTableStatement | truncateTableStatement | alterStatement | descStatement | showStatement | metastoreCheck | createViewStatement | dropViewStatement | createFunctionStatement | createMacroStatement | createIndexStatement | dropIndexStatement | dropFunctionStatement | dropMacroStatement | analyzeStatement | lockStatement | unlockStatement | lockDatabase | unlockDatabase | createRoleStatement | dropRoleStatement | grantPrivileges | revokePrivileges | showGrants | showRoleGrants | showRolePrincipals | showRoles | grantRole | revokeRole | setRole | showCurrentRole );])
        at org.antlr.runtime.DFA.noViableAlt(DFA.java:158)
        at org.antlr.runtime.DFA.predict(DFA.java:144)
        at org.apache.hadoop.hive.ql.parse.HiveParser.ddlStatement(HiveParser.java:2084)
        at org.apache.hadoop.hive.ql.parse.HiveParser.execStatement(HiveParser.java:1392)
        at org.apache.hadoop.hive.ql.parse.HiveParser.statement(HiveParser.java:1030)
        at org.apache.hadoop.hive.ql.parse.ParseDriver.parse(ParseDriver.java:199)
        at org.apache.hadoop.hive.ql.parse.ParseDriver.parse(ParseDriver.java:166)
        at org.apache.hadoop.hive.ql.Driver.compile(Driver.java:417)
        at org.apache.hadoop.hive.ql.Driver.compile(Driver.java:335)
        at org.apache.hadoop.hive.ql.Driver.compileInternal(Driver.java:1026)
        at org.apache.hadoop.hive.ql.Driver.runInternal(Driver.java:1091)
        at org.apache.hadoop.hive.ql.Driver.run(Driver.java:962)
        at org.apache.hadoop.hive.ql.Driver.run(Driver.java:952)
        at org.apache.hadoop.hive.cli.CliDriver.processLocalCmd(CliDriver.java:269)
        at org.apache.hadoop.hive.cli.CliDriver.processCmd(CliDriver.java:221)
        at org.apache.hadoop.hive.cli.CliDriver.processLine(CliDriver.java:431)
        at org.apache.hadoop.hive.cli.CliDriver.processLine(CliDriver.java:367)
        at org.apache.hadoop.hive.cli.CliDriver.executeDriver(CliDriver.java:750)
        at org.apache.hadoop.hive.cli.CliDriver.run(CliDriver.java:694)
        at org.apache.hadoop.hive.cli.CliDriver.main(CliDriver.java:633)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.apache.hadoop.util.RunJar.main(RunJar.java:212)
FAILED: ParseException line 1:4 cannot recognize input near 'DROP' '<EOF>' '<EOF>' in ddl statement
/opt/Big-Data-Benchmark-for-Big-Bench/engines/hive/conf/engineSettings.conf: line 13: DROP: command not found
cleaning dir /user/root/benchmarks/bigbench/queryResults/q01_hive_RUN_QUERY_0_result
Deleted /user/root/benchmarks/bigbench/queryResults/q01_hive_RUN_QUERY_0_result
cleaning dir /user/root/benchmarks/bigbench/temp/q01_hive_RUN_QUERY_0_temp
Deleted /user/root/benchmarks/bigbench/temp/q01_hive_RUN_QUERY_0_temp
cleaning log /opt/Big-Data-Benchmark-for-Big-Bench/logs/q01_hive_RUN_QUERY_0.log

jameszhouyi commented:
Hi，
Based on  joshuayao  finding, I think we need to re-evaluate the impact of applying 'eval' and so now is it possible to rollback this PR ?

Thanks

Manuel-Danisch commented:
Hi,

I reverted the pull request for now. I also tried the first suggested solution (echo ... | xargs), but it did not work either.

Best regards,

Manuel

Manuel-Danisch commented:
Hi,

in my opinion, the only way to get this working in a sane manner is switching BINARY_PARAMS from a simple string to a bash array. I implemented the changes in the new double-quote-fix branch and tested it with query 4. On our systems, it worked as expected.

Could you please checkout this branch, apply your changes and test if your double quoted string works with it?

Best regards,

Manuel

jameszhouyi commented:
Hi @Manuel-Danisch,
Thanks a lot for your efforts! May need a little more time to validate this new solution for us.

Thanks!

Manuel-Danisch commented:
Hi @jameszhouyi,

Do you have any updates on this issue?

Best regards,

Manuel

joshuayao commented:
Hi @Manuel-Danisch ,

We need more validations with Hive on Spark engine. After that, we'll provide feedback. Thanks.

Best Regards,
Yi

xhao1 commented:
Hi @Manuel-Danisch ,
For Hive on Spark, we finished some sanity validation and seems it works OK. (runBenchmark, runBenchmark with -c option. runQuery, cleanQuery, cleanQueries etc.). 
One small issue encountered seems not related to this PR30 change (please see issue 34 for details).
Thanks.

Best regards,
Xin

jameszhouyi commented:
Hi @Manuel-Danisch ,
It's OK for Spark SQL to pass JVM parameters in engineLocalSettings.conf. Thank you !

BR,
Yi
